projects/span_finder/scripts/preprocessing/preprocess_genia.py

```diff
@@ -0,0 +1,80 @@
+from pathlib import Path
+from typing import List
+
+import typer
+from spacy.tokens import Doc, DocBin, SpanGroup
+from spacy.training.converters import conll_ner_to_docs
+from wasabi import msg
+
+DOC_DELIMITER = "-DOCSTART- -X- O O\n"
+
+
+def parse_genia(
+    data: str, span_key: str, num_levels: int = 4, doc_delimiter: str = DOC_DELIMITER
+) -> List[Doc]:
+    """Parse GENIA dataset into spaCy docs
+
+    Our strategy here is to reuse the conll -> ner method from
+    spaCy and re-apply that n times. We don't want to write our
+    own ConLL/IOB parser.
+
+    Parameters
+    ----------
+    data: str
+        The raw string input as read from the IOB file
+    num_levels: int, default is 4
+        Represents how many times a label has been nested. In
+        GENIA, a label was nested four times at maximum.
+
+    Returns
+    -------
+    List[Doc]
+    """
+    docs = data.split("\n\n")
+    iob_per_level = []
+    for level in range(num_levels):
+        doc_list = []
+        for doc in docs:
+            tokens = [t for t in doc.split("\n") if t]
+            token_list = []
+            for token in tokens:
+                annot = token.split("\t")
+                # First element is always the token text
+                text = annot[0]
+                label = annot[level + 1]
+                _token = " ".join([text, label])
+                token_list.append(_token)
+            doc_list.append("\n".join(token_list))
+        annotations = doc_delimiter.join(doc_list)
+        iob_per_level.append(annotations)
+
+    # We then copy all the entities from doc.ents into
+    # doc.spans later on. But first, let's have a "canonical" docs
+    # to copy into
+    docs_per_level = [list(conll_ner_to_docs(iob)) for iob in iob_per_level]
+    docs_with_spans: List[Doc] = []
+
+    for docs in zip(*docs_per_level):
+        spans = [ent for doc in docs for ent in doc.ents]
+        doc = docs[0]
```

Comment:
Hi @svlandeg @thomashacker (I recall seeing a notif that asks about this line of code, but I can't seem to find it now so I'll add it here). So the way the GENIA IOB was originally constructed is that it has 4 "levels" due to their overlapping nature. What I did was create a `Doc` with the entities per each level (using `conll_ner_to_docs`)

However, we want all entities in a single `Doc`, so the way I "merged" them is to put all the entities into the `.spans` of the first `Doc` (arbitrary choice).